lib/gc/Dialect/Linalgx/Utils.cpp

```diff
@@ -0,0 +1,451 @@
+  if (!validBatch)
+    return false;
+
+  // check packing axis
+  auto getBatchAxisSet = [](llvm::SmallSet<int64_t, 8> &indexSet,
+                            ArrayRef<int64_t> batchDims) {
+    indexSet.insert(batchDims.begin(), batchDims.end());
+  };
+  auto getPackingAxisSet = [](ArrayRef<PackingMap> mapArray,
+                              llvm::SmallSet<int64_t, 8> &firstIndexSet,
+                              llvm::SmallSet<int64_t, 8> &secondIndexSet) {
+    for (auto &packingMap : mapArray) {
+      auto firstDims = packingMap.getFirst();
+      firstIndexSet.insert(firstDims.begin(), firstDims.end());
+      auto secondDims = packingMap.getSecond();
+      secondIndexSet.insert(secondDims.begin(), secondDims.end());
+    }
+  };
+  llvm::SmallSet<int64_t, 8> indexSetA;
+  llvm::SmallSet<int64_t, 8> indexSetB;
+  llvm::SmallSet<int64_t, 8> indexSetC;
+  getBatchAxisSet(indexSetA, attr.batchDimMap.getBatchA());
+  getBatchAxisSet(indexSetB, attr.batchDimMap.getBatchB());
+  getBatchAxisSet(indexSetC, attr.batchDimMap.getBatchC());
+  getPackingAxisSet(attr.mPacking, indexSetA, indexSetC);
+  getPackingAxisSet(attr.nPacking, indexSetB, indexSetC);
+  getPackingAxisSet(attr.kPacking, indexSetA, indexSetB);
+  bool checkAxis = (shapeA.getRank() == (int64_t)indexSetA.size()) &&
+                   (shapeB.getRank() == (int64_t)indexSetB.size()) &&
+                   (shapeC.getRank() == (int64_t)indexSetC.size());
+  if (!checkAxis)
+    return false;
+
+  // check packing dims match
+  auto matchBatch = [&](const BatchDimMap &batchDimMap) {
+    bool matchBatch = true;
+    for (int64_t i = 0; i < batchDimMap.getBatchNum(); i++) {
+      auto dimA = batchDimMap.getBatchA()[i];
+      auto dimB = batchDimMap.getBatchB()[i];
+      matchBatch = matchBatch && //
+                   (shapeA.getDimSize(dimA) == shapeB.getDimSize(dimB));
+      if (batchDimMap.isBatchMatmul()) {
+        auto dimC = batchDimMap.getBatchC()[i];
+        matchBatch = matchBatch && //
+                     (shapeA.getDimSize(dimA) == shapeC.getDimSize(dimC));
+      }
+    }
+    return matchBatch;
+  };
+  auto matchDims = [](ArrayRef<PackingMap> mapArray, ShapedType firstShape,
+                      ShapedType secondShape) {
+    for (auto &packingMap : mapArray) {
+      bool isDynamic = false;
+      int64_t firstSize = 1;
+      auto firstDims = packingMap.getFirst();
+      for (auto dim : firstDims) {
+        auto size = firstShape.getDimSize(dim);
+        if (size == ShapedType::kDynamic)
+          isDynamic = true;
+        firstSize *= size;
+      }
+      int64_t secondSize = 1;
+      auto secondDims = packingMap.getSecond();
+      for (auto dim : secondDims) {
+        auto size = secondShape.getDimSize(dim);
+        if (size == ShapedType::kDynamic)
+          isDynamic = true;
+        secondSize *= size;
+      }
+      if (isDynamic)
+        return false; // does not support dynamic dims
+      if (firstSize != secondSize)
```

Comment:
Equal size does not necessarily represent the same semantic. Shall we check AffineMap Expr instead?